### PR TITLE
Feat/manage third nav elems

### PR DESCRIPTION
**Pull request by kwajiehao**
**Note:** This PR needs to be reviewed together with PR #[71](https://github.com/isomerpages/isomercms-backend/pull/71) in the backend.

**Preface:** While writing this PR, I realized that a lot of the complex logic involved here stems from two things:
1. From the Workspace, we are able to create a collection page
2. Within a collection page, we can use the `ComponentSettingsModal` to reassign the page's third nav section.

In performing these actions, complex logic is involved in creating the file name to respect the Isomer naming conventions. And the fact that we are deleting and creating files while renaming them might suggest that we don't want to allow users to modify `Category` and third nav from the `ComponentSettingsModal`. Rather, it might make more sense to put these functionalities within the page-moving modal which Alex is working on. This would allow us to simplify the logic of the `ComponentSettingsModal`, which is currently very complex.

Alex, when you read this, please hit me up to discuss

# Overview

This PR follows up from PR #206 and resolves the rest of issue #202.

This PR's main role is to complete the functionality of the `ComponentSettingsModal` with regards to the page hierarchy layouts implemented in PR #206. This entails:
- Page creation (from the `Workspace` as well as from within a specific collection)
- File modification. We need to ensure that we can assign pages within collections to third nav sections, or create new third nav sections and assign pages to those sections

This turned out to be more complicated than expected, so I'll break the contents of this PR down into 3 parts:
1. Refactoring `ComponentSettingsModal`
2. Implementing a dropdown for the third nav title field
3. Coming up with file-naming logic when creating a new page within a collection / reassigning pages to and from third nav sections

## Refactoring `ComponentSettingsModal`

There were a few reasons why I felt a refactor was required:
1. There was confusion over the `category` variable, which sometimes referred to the state variable, and sometimes referred to the prop. To make things more confusing, the original `category` prop was also stored as a state variable called `originalCategory`.
2. `type` and `originalCategory` props were set as state variables even though they were never modified. They should remain as props and remain accessible as props.
3. The current team is more familiar working with hooks, which also simplifies the definition of functions (no need to rely on `this` keyword), and also on the usage of props and state (removes the need to always destructure from props or `this.state`)

## Implementing third nav title dropdown

I attempted to use `CreatableSelect`, the component used to render category dropdown, to also render the third nav title options. The problem with this is that the API call to retrieve the third nav titles takes quite a bit of time. Even though I set the dropdown options as a state hook, the dropdown did not re-render the component with the full third nav title options after they were retrieved. I had to close and reopen the `ComponentSettingsModal` for the new options to be displayed.

Therefore, I chose to switch to use its asynchronous version, `AsyncCreatableSelect`, which allows you to pass in a promise which resolves to the dropdown options. This is perfect since we can just pass in the async function containing the API call to retrieve third nav titles.

It's important to note that state is stored on the parent component, `CollectionPagesSection`, instead of the `ComponentSettingsModal` itself. This is so that the third nav title dropdown options can be reused and passed as props when other pages in the collection need it, instead of making the same API call again to retrieve the options.

## File-naming logic

This section is the most complex. This is necessarily so because the file-naming convention for Isomer is complex.

### Isomer file-naming convention

Example pages in a collection:
```
1-test.md
2a-first-third-nav.md
2b-first-third-nav-b.md
3-test-b.md
4a-second-third-nav.md
4b-second-third-nav-b.md
```

In such a collection, the alphanumeric prefix, such as `1` or `2a`, is known as the **group identifier**. Pages which are not in a third nav have a purely numeric group identifier, whereas pages in a third nav have an alphanumeric group identifier. Pages in the same third nav **must share** the same number, with only the alphabet being incremented (i.e. `2a`, `2b`, `2c`, etc.). If two pages which have the same third nav title are not alphanumerically adjacent (i.e. `2a-<title>.md` and `2b-<title>.md` are adjacent, but `2a-<title>.md` and `4b-<title>.md` are not), then the Isomer template interprets it as two separate third nav sections.

Screenshot 2020-11-12 at 2 37 11 AM

Therefore, when we are creating a new file to be inserted into the collection, moving a non-third-nav-collection page into a third nav section, etc. we need to be mindful of how we name the new file, or the left navigation bar of the page will be jumbled up.

### File-naming logic

Since we have decided not to deal with navigation before the Nov/Dec release, the strategy here is simple: to move the new file to the bottom of its new list.

For example, if we're moving a page into a third nav, we want it to be the last element in that third nav section. If we reuse the example collection above, if I want to move `4b-second-third-nav-b.md` into the first third nav section, it would be named `2c-second-third-nav-b.md`.

```
1-test.md
2a-first-third-nav.md
2b-first-third-nav-b.md
2c-second-third-nav-b.md
3-test-b.md
4a-second-third-nav.md
```

And if we're moving a page into a collection, then we want it to be the last element in that collection. If I created a new page called `New Page` in this same example collection, the new collection structure would be:

```
1-test.md
2a-first-third-nav.md
2b-first-third-nav-b.md
2c-second-third-nav-b.md
3-test-b.md
4a-second-third-nav.md
5-new-page.md
```

The new name respects the Isomer file-naming convention by incrementing the group identifier of the new file in an alphanumeric manner.

### Test cases to check for

**From workspace**
1. Create new collection page NOT IN third nav from workspace
2. Create new collection page IN third nav from workspace

**From within a collection**
3. Create new collection page from within a collection
- inside a third nav section
- regular collection page

4. Change a third nav page to a regular page 
5. Change a regular page to a third nav page
6. Move a page that is inside a third nav section into another third nav section 

**Creating a new third nav section**
7. Create a new third nav section and assign a regular page in the collection to it

### Others

Please refer to relevant commit message for more details.

### CODE REVIEW

**Review by alexanderleegs — CHANGES_REQUESTED** [review]
Left comments.

In addition to these, there were also some issues I noticed:
- I can't seem to create a new 3rd nav from the workspace page - it just attempts to load
- In contrast, when inside a collection page, creating a new 3rd nav works, but the search functionality doesn't seem to be working - the options are not filtered as the user continues typing
- When creating a file in a new 3rd nav, the file is named with only the integer (e.g. 0-title), which causes subsequent items when placed into the 3rd nav to not follow this naming convention (e.g. the next one would be called 1-title), which causes issues with the display

File: src/components/ComponentSettingsModal.jsx [hunk 1]
```diff
@@ -20,393 +22,451 @@ import DeleteWarningModal from './DeleteWarningModal';
...
-                    value={thirdNavTitle}
-                    onFieldChange={this.changeHandler}
+                    title="Permalink"
+                    id="permalink"
+                    value={permalink ? permalink : ''}
+                    errorMessage={errors.permalink}
+                    isRequired={true}
+                    onFieldChange={changeHandler}
+                    disabled={!isPost}
+                    fixedMessage={originalCategory ? `/${originalCategory}/${thirdNavTitle ? `${thirdNavTitle}/` : ''}` : '/'}
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/components/ComponentSettingsModal.jsx]: Fixed with [05132de](https://github.com/isomerpages/isomercms-frontend/pull/218/commits/05132de418769944988af0e865cfd04720ed4bcc)

File: src/components/ComponentSettingsModal.jsx [hunk 2]
```diff
@@ -20,393 +22,451 @@ import DeleteWarningModal from './DeleteWarningModal';
...
+
+        const baseUrl = `${process.env.REACT_APP_BACKEND_URL}/sites/${siteName}${originalCategory ? type === "resource" ? `/resources/${originalCategory}` : `/collections/${originalCategory}` : ''}`
+        setBaseApiUrl(baseUrl)
+
+        const fetchData = async () => {
+            // Retrieve the list of all page/resource categories for use in the dropdown options. Also sets the default category if none is specified.
+            if (type === 'resource') {
+                const resourcesResp = await axios.get(`${process.env.REACT_APP_BACKEND_URL}/sites/${siteName}/resources`);
+                const { resources } = resourcesResp.data;
+                setAllCategories(resources.map((category) => category.dirName))
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/components/ComponentSettingsModal.jsx]: Fixed with [ceafd41](https://github.com/isomerpages/isomercms-frontend/pull/218/commits/ceafd41c8f8b48f3e2f0832c8fa2429811111e90)

File: src/utils.js [hunk 3]
```diff
@@ -270,4 +293,135 @@ export async function saveFileAndRetrieveUrl(fileInfo) {
...
+      const apiUrl = `${process.env.REACT_APP_BACKEND_URL}/sites/${siteName}/collections/${category}`
+      const groupIdentifier = await generateGroupIdentifier(null, false, apiUrl)
+      newFileName = generateCollectionPageFileName(title, groupIdentifier);
+
+    // Case: Creating a new page from the workspace and assigning to collection AND third nav
+    } else if (!collectionPageData && thirdNavTitle) {
+      const apiUrl = `${process.env.REACT_APP_BACKEND_URL}/sites/${siteName}/collections/${category}/pages`
+      const groupIdentifier = await generateGroupIdentifier(null, true, apiUrl, false, thirdNavTitle)
+      newFileName = generateCollectionPageFileName(title, groupIdentifier);
+    }
```
Comment (alexanderleegs) [inline src/utils.js:352]: These cases don't cover the generation of new 3rd navs from within a collection/from the workspace

Comment (kwajiehao) [reply src/utils.js]: Fixed with [fba6ecf](https://github.com/isomerpages/isomercms-frontend/pull/218/commits/fba6ecf9e7b9991448239a69b7746405ab55bf9f)

**Review by alexanderleegs — APPROVED** [review]
lgtm

**Pull request merged**